### ISSUE TITLE
Enabling http compression results in empty http request body

### ISSUE THREAD

**Issue by kpstolk** [elastic/elasticsearch-net#3913]
**NEST/Elasticsearch.Net version**:
7.0.0

**Elasticsearch version**:
7.2.0 & 6.8.0

**Description of the problem including expected versus actual behavior**:
Actual: Enabling `EnableHttpCompression()` does not send the `PostData` in the http request body.
Expected: `PostData` to be sent in the http body of the request.

**Steps to reproduce**:
 1. Create two `LowLevel` clients, one without http compression and one with `EnableHttpCompression()` configured.

```
// Without http compression:
var settings = new ConnectionSettings(new Uri("https://localhost.fiddler:9200"))
    .BasicAuthentication("elastic", "elastic");
var client = new ElasticClient(settings);

// With http compression:
settings.EnableHttpCompression();
var client2 = new ElasticClient(settings);
```

2. Send a request including `PostData` for both clients:

```
var response = await client.LowLevel.DoRequestAsync<DynamicResponse>(
    HttpMethod.PUT,
    "my-index1",
    CancellationToken.None,
    PostData.String("{\"aliases\": { \"my-alias1\": { } } }")
);

var response2 = await client2.LowLevel.DoRequestAsync<DynamicResponse>(
    HttpMethod.PUT,
    "my-index2",
    CancellationToken.None,
    PostData.String("{\"aliases\": { \"my-alias2\": { } } }")
);
```

3. Fiddler shows both http requests with and without a body:

```
PUT https://localhost:9200/my-index1 HTTP/1.1
Accept: application/json
User-Agent: elasticsearch-net/7.0.0 (Microsoft Windows 10.0.17134 ; .NET Core 4.6.27617.04; Nest)
Authorization: Basic ZWxhc3RpYzplbGFzdGlj
Content-Type: application/json
Host: localhost:9200
Content-Length: 34

{"aliases": { "my-alias1": { } } }


PUT https://localhost:9200/my-index2 HTTP/1.1
Accept: application/json
User-Agent: elasticsearch-net/7.0.0 (Microsoft Windows 10.0.17134 ; .NET Core 4.6.27617.04; Nest)
Accept-Encoding: gzip, deflate, gzip, deflate
Authorization: Basic ZWxhc3RpYzplbGFzdGlj
Content-Type: application/json
Host: localhost:9200
Content-Length: 0
```

 4. In Elasticsearch `GET my-index*/_alias` returns both indices but the request with http compression enabled obviously doesn't have the requested alias attached:

```
{
  "my-index2" : {
    "aliases" : { }
  },
  "my-index1" : {
    "aliases" : {
      "my-alias1" : { }
    }
  }
}
```

**Comment by russcam** [comment]
> NEST/Elasticsearch.Net version:
> 7.0.0
> 
> Elasticsearch version:
> 6.8.0

[NEST 7.0.0 is not compatible with 6.8.0](https://github.com/elastic/elasticsearch-net#compatibility-matrix); please use the latest NEST 6.x package with Elasticsearch 6.8.0.

----

I don't see this issue with 7.0.0. Here's an example, with [Fiddler as a proxy](https://www.elastic.co/guide/en/elasticsearch/client/net-api/current/logging-with-fiddler.html)

```c#
private static void Main()
{
	MainAsync().Wait();
}

private static async Task MainAsync()
{
	var defaultIndex = "posts";
	var pool = new SingleNodeConnectionPool(new Uri("http://ipv4.fiddler:9200"));

	var settings = new ConnectionSettings(pool)
		.DefaultIndex(defaultIndex);

	var client = new ElasticClient(settings);

	var response = await client.LowLevel.DoRequestAsync<DynamicResponse>(
		HttpMethod.PUT, 
		"my-index", 
		CancellationToken.None, 
		PostData.String("{\"aliases\": { \"my-alias\": { } } }")
	);
}
```

The request in Fiddler

```
PUT http://127.0.0.1:9200/my-index HTTP/1.1
Accept: application/json
Content-Type: application/json
User-Agent: elasticsearch-net/7.0.0 (Microsoft Windows 10.0.17763; .NET Framework 4.7.3416.0; Nest)
Host: 127.0.0.1:9200
Content-Length: 33

{"aliases": { "my-alias": { } } }
```

and response

```
HTTP/1.1 502 Fiddler - Connection Failed
Date: Thu, 04 Jul 2019 00:22:11 GMT
Content-Type: text/html; charset=UTF-8
Connection: close
Cache-Control: no-cache, must-revalidate
Timestamp: 10:22:11.712

[Fiddler] The connection to '127.0.0.1' failed. <br />Error: ConnectionRefused (0x274d). <br />System.Net.Sockets.SocketException No connection could be made because the target machine actively refused it 127.0.0.1:9200                                                                                                                                                                                                                                                                                                     
````

![image](https://user-images.githubusercontent.com/208231/60632053-70454d00-9e46-11e9-8948-2349f1e2bd94.png)

_(This has highlighted another issue where the serializer attempts to deserialize the response and throws a `JsonParsingException`, which I'll open another issue for)_.

Is it possible that the index already exists before trying to create it again with an alias? If it does already exist, the call will fail with a 400 HTTP response code and the body will contain an error object

```
HTTP/1.1 400 Bad Request
content-type: application/json; charset=UTF-8
content-length: 373

{"error":{"root_cause":[{"type":"resource_already_exists_exception","reason":"index [my-index/Kc_kgTBOQjyakZFFAiFJww] already exists","index_uuid":"Kc_kgTBOQjyakZFFAiFJww","index":"my-index"}],"type":"resource_already_exists_exception","reason":"index [my-index/Kc_kgTBOQjyakZFFAiFJww] already exists","index_uuid":"Kc_kgTBOQjyakZFFAiFJww","index":"my-index"},"status":400}
```

which can be checked with

```c#
if (!response.Success)
{
    Console.WriteLine(response.Body["error"]["reason"].ToString());
}
```

**Comment by kpstolk** [comment]
Thanks for your investigation, I tried with Elasticsearch 7.x and now it works as expected.
I didn't think the Elasticsearch version would be the issue here because the request itself didn't include the body to the server in the first place.

I didn't know that NEST 7.x is not backwards compatible, is there a solution to workaround this? We can't upgrade our applications and Elasticsearch at the same time.
The first issue I hit was the `hits.total` response which has changed, but [this particular issue](https://github.com/elastic/elasticsearch-net/issues/3791#issuecomment-498099262) is compatible between Elasticsearch 6.x and 7.x?

**Comment by kpstolk** [comment]
It seems I'm wrong about the Elasticsearch 7.x assumption, this has actually nothing to do with it I just found out.
I had a minor difference in the way I configured the client while testing against Elasticsearch 7.2 which actually seems to be the issue.
For my first test with Elasticsearch 6.8 I had `EnableHttpCompression()` configured while with the Elasticsearch 7.2 test I had not.

I have updated the issue description with this information.
Does NEST 7.0 handle http compression differently than in NEST 6.8?

**Comment by Mpdreamz** [comment]
Thank you @kpstolk this is indeed the case sadly.

Nest 7.x did change the stream handling on `.NET Core` where we moved to a non buffering `HttpContent` implementation.

I've opened: https://github.com/elastic/elasticsearch-net/pull/3921

Including measures to make sure this does not slip again in the future.

**Comment by kpstolk** [comment]
Thanks @Mpdreamz for confirming this issue.

Regarding compatibility with NEST and Elasticsearch, is there any guidance on how to deal with that?

> I didn't know that NEST 7.x is not backwards compatible, is there a solution to workaround this? We can't upgrade our applications and Elasticsearch at the same time.
The first issue I hit was the `hits.total` response which has changed, but [this particular issue](https://github.com/elastic/elasticsearch-net/issues/3791#issuecomment-498099262) is compatible between Elasticsearch 6.x and 7.x?

**Comment by russcam** [comment]
@kpstolk Thank you for further investigation.

With regards to compatibility, we have been releasing versioned clients to our [CI package repository](https://ci.appveyor.com/nuget/elasticsearch-net). A versioned client is IL-rewritten such that the assembly name and all namespaces are changed, allowing a versioned client to be referenced within a project in addition to an official release client package from nuget.

For example, with a CI nuget package source configured in a Nuget.config file

```xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <packageSources>
    <add key="Appveyor CI" value="https://ci.appveyor.com/nuget/elasticsearch-net" />
  </packageSources>
</configuration>
```

A project can pull in the latest `Nest.v6` package from CI, in addition to `Nest` package version 7.0.0 from Nuget. For example,

```xml
<Project Sdk="Microsoft.NET.Sdk">
    <PropertyGroup>
        <OutputType>Exe</OutputType>
        <TargetFramework>netcoreapp2.2</TargetFramework>
    </PropertyGroup>
    <ItemGroup>
      <PackageReference Include="NEST" Version="7.0.0" />
      <PackageReference Include="NEST.v6" Version="6.9.0-ci20190620T132817" />
    </ItemGroup>
</Project>
```

Then, both can be used in the project side by side

```c#
public static void Main()
{
    // 6.x client from Nest.v6 package
    var settings6x = new Nest6.ConnectionSettings(new Elasticsearch.Net6.SingleNodeConnectionPool(uri))
        .DefaultIndex("posts");

    var client6x = new Nest6.ElasticClient(settings6x);

    // 7.x client from Nest 7.0.0 package
    var settings = new ConnectionSettings(new SingleNodeConnectionPool(uri))
        .DefaultIndex("posts");

    var client = new ElasticClient(settings);
}
```

and custom logic can be implemented to determine which client to use.

The intention with providing these is to help in the short-term in upgrading the client to the next major version, at the same time as upgrading Elasticsearch. I think this would be a good topic to include in documentation and a blog post 😃

**Comment by russcam** [comment]
@kpstolk This is now [fixed in NEST 7.0.1](https://www.nuget.org/packages/NEST/7.0.1) 👍

**Comment by kpstolk** [comment]
@russcam Thanks for releasing the new version, that was quick! 😄 
I'll take a look at the `Nest.v6` package and see if that works for me.
